storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/scheduling/ConstraintSolverStrategy.java

```diff
@@ -45,64 +45,142 @@
     //hard coded max number of states to search
     private static final Logger LOG = LoggerFactory.getLogger(ConstraintSolverStrategy.class);
 
-    //constraints and spreads
-    private Map<String, Map<String, Integer>> constraintMatrix;
-    private HashSet<String> spreadComps = new HashSet<>();
+    /**
+     * Component constraint as derived from configuration.
+     * This is backward compatible and can parse old style Config.TOPOLOGY_RAS_CONSTRAINTS and Config.TOPOLOGY_SPREAD_COMPONENTS.
+     * New style Config.TOPOLOGY_RAS_CONSTRAINTS is map where each component has a list of other incompatible components
+     * and an optional number that specifies the maximum co-location count for the component on a node.
+     *      { "comp-1": [ 2, "comp-2", "comp-3" ], # comp-1 cannot exist on same node as comp-2 or comp-3, and at most 2 comp-1 same node
```

Comment:
Please correct the comment, @Ethanlm  thinks the component exclusion is at worker level

Reply:
The description can be modified to https://github.com/apache/storm/blob/master/storm-client/src/jvm/org/apache/storm/Config.java#L307-L312
Also, validator needs to allow for list or map.
The configuration could be more explicit in mentioning node level collation and component isolation.
`{"comp-1": { "maxNodeCollations" : 2, "incompatibleComponents": [ "comp-2", "comp-3"  ]} }`

Reply:
Updated Config.java, ConfigValidationAnnotations.java, ConfigValidations.java, ConstraintSolverStrategy and TestConstraintSolverStrategy.